### DISCUSSION TITLE
Routing with dynamic cluster?

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/yarp#2632] [Q&A] [answered]
I would like to define a route as `/p/{clusterName}/{**any}` and forward it to `clusterName` possibly applying some transformers. 

Note any valid "clusterName" here would only have a single destination.

I think there's a few ways to do this: custom transform and setting `DestinationPrefix` or `proxyrequest.RequestUri` or even a custom load balancer or using `ReassignProxyRequest`, etc. 

Which one is the preferred way to do this?

**Comment by MihaZupan** [comment, marked as answer]
I'd say the preferred approach depends on how many such clusters you have / whether they're dynamic (you add/remove them while the proxy is running).

If you only have a handful of static ones, you could simply define them all in the configuration file.

If there are a lot of different ones / they're updated often, it may be easiest to switch to a code-based approach. Since you mentioned only having one destination per cluster, you're unlikely to benefit from features such as session affinity, load balancing, or health checks, so you could look at using [direct forwarding](https://microsoft.github.io/reverse-proxy/articles/direct-forwarding.html).
E.g. something along the lines of
```c#
var destinations = new Dictionary<string, string>
{
    { "foo", "https://foo.com" }
};

app.Map("/p/{clusterName}/{**remainder}", async (HttpContext context, IHttpForwarder forwarder,
    [FromRoute] string clusterName,
    [FromRoute] string remainder) =>
{
    if (destinations.TryGetValue(clusterName, out var destination))
    {
        context.Request.Path = $"/{remainder}";
        await forwarder.SendAsync(context, destination, client);
    }
});
```

**Reply by mdutton3** [reply]
It looks like the link to Direct Forwarding can now be found at:
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/yarp/direct-forwarding
